File: Platform/Workspace/Typescript/workspace/libs/workspace/data/src/Equals.ts
import { ObjectType, PropertyType } from '@allors/workspace/meta';
import { ISessionObject, CompositeTypes, UnitTypes,serialize } from '@allors/workspace/domain';

import { ParameterizablePredicateArgs, ParameterizablePredicate } from './ParameterizablePredicate';

export interface EqualsArgs extends ParameterizablePredicateArgs, Pick<Equals, 'propertyType' | 'value' | 'object'> {}

export class Equals extends ParameterizablePredicate {
  public propertyType: PropertyType;
  public value?: UnitTypes;
  public object?: CompositeTypes;

  constructor(propertyType: PropertyType);
  constructor(args: EqualsArgs);
  constructor(args: EqualsArgs | PropertyType) {
    super();

    if ((args as PropertyType).objectType) {
      this.propertyType = args as PropertyType;
    } else {
      Object.assign(this, args);
      this.propertyType = (args as EqualsArgs).propertyType;
    }
  }

  get objectType(): ObjectType {
    return this.propertyType.objectType;
  }

  public toJSON(): any {
    return {
      kind: 'Equals',
      dependencies: this.dependencies,
      propertytype: this.propertyType.id,
      parameter: this.parameter,
      value: serialize(this.value),
      object: this.object && (this.object as ISessionObject).id ? (this.object as ISessionObject).id : this.object,
    };
  }
}

File: dotnet/platform/workspace/csharp/allors.workspace/data/equals.cs
// <copyright file="Equals.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;
    using Allors.Protocol.Data;
    using Allors.Workspace.Meta;

    public class Equals : IPropertyPredicate
    {
        public string[] Dependencies { get; set; }

        public Equals(IPropertyType propertyType = null) => this.PropertyType = propertyType;

        /// <inheritdoc/>
        public IPropertyType PropertyType { get; set; }

        public ISessionObject Object { get; set; }

        public object Value { get; set; }

        public string Parameter { get; set; }

        public Predicate ToJson() =>
            new Predicate
            {
                kind = PredicateKind.Equals,
                dependencies = this.Dependencies,
                propertyType = this.PropertyType.Id,
                @object = this.Object?.Id.ToString(),
                value = UnitConvert.ToString(this.Value),
                parameter = this.Parameter,
            };
    }
}
